test/unit/mc/test_triangular_solve_local.cpp

```diff
@@ -0,0 +1,337 @@
+  //      = (res_ij * op(a)_jj + (kk-1) * gamma) / alpha,
+  // where gamma = (j+1) / (i+2) * exp(I*(i+2*j)),
+  //       kk = j+1 if op(a) is an upper triangular matrix, or
+  //       kk = n-j if op(a) is an lower triangular matrix.
+  // Therefore
+  // b_ij = (res_ij + (kk-1) * gamma) / alpha, if diag == Unit
+  // b_ij = kk * gamma / alpha, otherwise.
+
+  bool op_a_lower = false;
+  if ((uplo == blas::Uplo::Lower && op == blas::Op::NoTrans) ||
+      (uplo == blas::Uplo::Upper && op != blas::Op::NoTrans))
+    op_a_lower = true;
+
+  auto res_b = [](const GlobalElementIndex& index) {
+    double i = index.row();
+    double k = index.col();
+
+    return TypeUtilities<T>::polar((k + .5) / (i + 2), i + k);
+  };
+
+  auto el_op_a = [op_a_lower, diag](const GlobalElementIndex& index) {
+    if ((op_a_lower && index.row() < index.col()) || (!op_a_lower && index.row() > index.col()) ||
+        (diag == blas::Diag::Unit && index.row() == index.col()))
+      return TypeUtilities<T>::element(-9.9, 0);
+
+    double k = index.row();
+    double j = index.col();
+
+    return TypeUtilities<T>::polar((j + 1) / (k + .5), 2 * j - k);
+  };
+
+  auto el_b = [n, alpha, diag, op_a_lower, res_b](const GlobalElementIndex& index) {
+    BaseType<T> kk = op_a_lower ? n - index.col() : index.col() + 1;
+
+    double i = index.row();
+    double j = index.col();
+    T gamma = TypeUtilities<T>::polar((j + 1) / (i + 2), i + 2 * j);
+
+    if (diag == blas::Diag::Unit)
+      return ((kk - 1) * gamma + res_b(index)) / alpha;
+    else
+      return kk * gamma / alpha;
+  };
+
+  return std::make_tuple(el_op_a, el_b, res_b);
+}
+
+template <class T>
+void testTriangularSolve(blas::Side side, blas::Uplo uplo, blas::Op op, blas::Diag diag, T alpha,
+                         Matrix<T, Device::CPU>& matA, Matrix<T, Device::CPU>& matB,
+                         Matrix<T, Device::CPU>& matX) {
+  std::function<T(const GlobalElementIndex&)> el_op_a, el_b, res_b;
+
+  auto m = matB.size().rows();
+  auto n = matB.size().cols();
+
+  if (side == blas::Side::Left)
+    std::tie(el_op_a, el_b, res_b) =
+        testTriangularSolveElementFunctionsLeft<T>(uplo, op, diag, alpha, m);
+  else
+    std::tie(el_op_a, el_b, res_b) =
+        testTriangularSolveElementFunctionsRight<T>(uplo, op, diag, alpha, n);
+
+  set(matA, el_op_a, op);
+  set(matB, el_b);
+  set(matX, res_b);
+
+  triangular_solve(side, uplo, op, diag, alpha, matA, matB);
+
+  CHECK_MATRIX_NEAR(res_b, matB, 20 * (matB.size().rows() + 1) * TypeUtilities<T>::error,
+                    20 * (matB.size().rows() + 1) * TypeUtilities<T>::error);
+}
+
+TYPED_TEST(TriangularSolveLocalTest, Correctness) {
+  for (auto diag : blas_diags) {
+    for (auto op : blas_ops) {
+      for (auto side : blas_sides) {
+        for (auto uplo : blas_uplos) {
+          for (const auto& size : square_sizes) {
+            for (const auto& block_size : square_block_sizes) {
+              for (const auto& col : col_b) {
```

Comment:
What about an approach like https://github.com/eth-cscs/DLA-Future/blob/master/test/unit/test_blas_tile.cpp#L177-L205 ?

Reply:
Something like this? https://github.com/eth-cscs/DLA-Future/pull/125/commits/b9d9dafb5112f06219d5aefb3e48392f712ed987#diff-f45b31048033a4fa8a62fa3d6ef83df2#L204-L216